File: lib/csharp/microsoft.bot.builder.skills/Microsoft.Bot.Builder.Skills/SkillWebSocketAdapter.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Net;
using System.Security.Claims;
using System.Threading;
using System.Threading.Tasks;
using Microsoft.AspNetCore.Http;
using Microsoft.Bot.Builder.Integration.AspNet.Core;
using Microsoft.Bot.Builder.Skills.Auth;
using Microsoft.Bot.Builder.Solutions;
using Microsoft.Bot.StreamingExtensions.Transport;
using Microsoft.Bot.StreamingExtensions.Transport.WebSockets;

namespace Microsoft.Bot.Builder.Skills
{
    /// <summary>
    /// This adapter is responsible for accepting a bot-to-bot call over websocket transport.
    /// It'll perform the following tasks:
    /// 1. Authentication.
    /// 2. Create RequestHandler to handle follow-up websocket frames.
    /// 3. Start listening on the websocket connection.
    /// </summary>
    public class SkillWebSocketAdapter : IBotFrameworkHttpAdapter
    {
        private readonly IBotTelemetryClient _botTelemetryClient;
        private readonly SkillWebSocketBotAdapter _skillWebSocketBotAdapter;
        private readonly IAuthenticator _authenticator;
        private readonly Stopwatch _stopWatch;

        public SkillWebSocketAdapter(
            SkillWebSocketBotAdapter skillWebSocketBotAdapter,
            BotSettingsBase botSettingsBase,
            IWhitelistAuthenticationProvider whitelistAuthenticationProvider,
            IBotTelemetryClient botTelemetryClient = null)
        {
            _skillWebSocketBotAdapter = skillWebSocketBotAdapter ?? throw new ArgumentNullException(nameof(skillWebSocketBotAdapter));
            if (botSettingsBase == null)
            {
                throw new ArgumentNullException(nameof(botSettingsBase));
            }

            if (whitelistAuthenticationProvider == null)
            {
                throw new ArgumentNullException(nameof(whitelistAuthenticationProvider));
            }

            var authenticationProvider = new MSJwtAuthenticationProvider(botSettingsBase.MicrosoftAppId);
            _authenticator = new Authenticator(authenticationProvider, whitelistAuthenticationProvider);

            _botTelemetryClient = botTelemetryClient ?? NullBotTelemetryClient.Instance;
            _stopWatch = new Stopwatch();
        }

        public async Task ProcessAsync(HttpRequest httpRequest, HttpResponse httpResponse, IBot bot, CancellationToken cancellationToken = default(CancellationToken))
        {
            if (httpRequest == null)
            {
                throw new ArgumentNullException(nameof(httpRequest));
            }

            if (httpResponse == null)
            {
                throw new ArgumentNullException(nameof(httpResponse));
            }

            if (bot == null)
            {
                throw new ArgumentNullException(nameof(bot));
            }

            if (!httpRequest.HttpContext.WebSockets.IsWebSocketRequest)
            {
                httpResponse.StatusCode = (int)HttpStatusCode.BadRequest;
                await httpResponse.WriteAsync("Upgrade to WebSocket required.", cancellationToken: cancellationToken).ConfigureAwait(false);
                return;
            }

            var claimsIdentity = await _authenticator.Authenticate(httpRequest, httpResponse).ConfigureAwait(false);

            await CreateWebSocketConnectionAsync(claimsIdentity, httpRequest.HttpContext, bot).ConfigureAwait(false);
        }

        private async Task CreateWebSocketConnectionAsync(ClaimsIdentity claimsIdentity, HttpContext httpContext, IBot bot)
        {
            var socket = await httpContext.WebSockets.AcceptWebSocketAsync().ConfigureAwait(false);
            var handler = new SkillWebSocketRequestHandler(claimsIdentity, _botTelemetryClient);
            var server = new WebSocketServer(socket, handler);
            server.Disconnected += Server_Disconnected;
            _skillWebSocketBotAdapter.Server = server;
            handler.Bot = bot;
            handler.SkillWebSocketBotAdapter = _skillWebSocketBotAdapter;

            _botTelemetryClient.TrackTrace("Starting listening on websocket", Severity.Information, null);
            _stopWatch.Start();
            var startListening = server.StartAsync();
            Task.WaitAll(startListening);
        }

        private void Server_Disconnected(object sender, DisconnectedEventArgs e)
        {
            if (_stopWatch.IsRunning)
            {
                _stopWatch.Stop();

                _botTelemetryClient.TrackEvent("SkillWebSocketOpenCloseLatency", null, new Dictionary<string, double>
                {
                    { "Latency", _stopWatch.ElapsedMilliseconds },
                });
            }
        }
    }
}

File: sdk/typescript/libraries/botbuilder-skills/src/websocket/skillWebSocketAdapter.ts
import { BotFrameworkAdapter, BotFrameworkAdapterSettings, BotTelemetryClient, NullTelemetryClient,
    TurnContext, WebRequest, WebResponse } from 'botbuilder';
import { Server, Socket } from 'microsoft-bot-protocol-websocket';
import { BotCallbackHandler } from '../activityHandler';
import { IAuthenticationProvider } from '../auth';
import { SkillWebSocketBotAdapter } from './skillWebSocketBotAdapter';
import { SkillWebSocketRequestHandler } from './skillWebSocketRequestHandler';

/**
 * This adapter is responsible for accepting a bot-to-bot call over websocket transport.
 * It'll perform the following tasks:
 * 1. Authentication.
 * 2. Create RequestHandler to handle follow-up websocket frames.
 * 3. Start listening on the websocket connection.
 */
export class SkillWebSocketAdapter extends BotFrameworkAdapter {
    private readonly authenticationProvider?: IAuthenticationProvider;
    private readonly telemetryClient: BotTelemetryClient;
    private readonly botAdapter: SkillWebSocketBotAdapter;

    public constructor(
        botAdapter: SkillWebSocketBotAdapter,
        authenticationProvider?: IAuthenticationProvider,
        telemetryClient?: BotTelemetryClient,
        config?: Partial<BotFrameworkAdapterSettings>
    ) {
        super(config);
        this.botAdapter = botAdapter;
        this.authenticationProvider = authenticationProvider;
        this.telemetryClient = telemetryClient || new NullTelemetryClient();
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/tslint/config
    public async processActivity(req: WebRequest, res: WebResponse, logic: (context: TurnContext) => Promise<any>): Promise<void> {
        await this.createWebSocketConnection(req, logic);
    }

    // eslint-disable-next-line @typescript-eslint/no-explicit-any, @typescript-eslint/tslint/config
    private async createWebSocketConnection(req: any, bot: BotCallbackHandler): Promise<void> {
        // MISSING found an equivalent to websocket in TypeScript
        // eslint-disable-next-line @typescript-eslint/tslint/config
        const socket: Socket = req.socket;
        const handler: SkillWebSocketRequestHandler = new SkillWebSocketRequestHandler(this.telemetryClient);
        const server: Server = new Server(socket, handler);
        // MISSING the Server class does not exposes Disconnected handler
        // in C# server.Disconnected += Server_Disconnected;
        this.botAdapter.server = server;
        handler.bot = bot;
        handler.activityHandler = this.botAdapter;

        await server.startAsync();
    }
}
